test/scripts/products/h22-validate-that-rate-limit-service-is-working-as-expected/h22-validate-that-rate-limit-service-is-working-as-expected.go

```diff
@@ -0,0 +1,504 @@
+
+		time.Sleep(time.Second)
+	}
+
+	if !overallSuccess {
+		fmt.Println("Test failed, not all iterations succeeded")
+		os.Exit(1)
+	}
+}
+
+// createRedis creates a Redis throwaway instance, and returns a channel that
+// will send the name of the Redis Pod once it's ready
+func createRedis(ctx context.Context, client k8sclient.Client, namespace string) (<-chan struct {
+	PodName string
+	Error   error
+}, error) {
+	redis := &crov1alpha1.Redis{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "throw-away-redis-pod",
+			Namespace: namespace,
+		},
+		Spec: crov1alpha1.RedisSpec{
+			SecretRef: &types.SecretRef{
+				Name: "throw-away-redis-ref",
+			},
+			Tier: "development",
+			Type: "workshop",
+		},
+	}
+
+	if err := client.Create(ctx, redis); err != nil {
+		return nil, err
+	}
+
+	done := make(chan struct {
+		PodName string
+		Error   error
+	})
+
+	go func() {
+		for {
+			if err := client.Get(ctx, k8sclient.ObjectKey{
+				Name:      redis.Name,
+				Namespace: redis.Namespace,
+			}, redis); err != nil {
+				done <- struct {
+					PodName string
+					Error   error
+				}{Error: err}
+				return
+			}
+
+			if redis.Status.Phase != "complete" {
+				fmt.Printf("  ☁️  Waiting for throwaway Redis container to complete. Current phase: %s...\n", redis.Status.Phase)
+				time.Sleep(5 * time.Second)
+				continue
+			}
+
+			podsList := &corev1.PodList{}
+			if err := client.List(ctx, podsList, &k8sclient.ListOptions{
+				Namespace: redis.Namespace,
+			}); err != nil {
+				done <- struct {
+					PodName string
+					Error   error
+				}{Error: err}
+				return
+			}
+
+			for _, pod := range podsList.Items {
+				if strings.HasPrefix(pod.Name, redis.Name) {
+					done <- struct {
+						PodName string
+						Error   error
+					}{PodName: pod.Name}
+					return
+				}
+			}
+
+			done <- struct {
+				PodName string
+				Error   error
+			}{Error: errors.New("Redis created but Pod not found")}
+			break
+		}
+	}()
+
+	return done, nil
+}
+
+// cleanupRedis deletes the throwaway Redis instance
+func cleanupRedis(ctx context.Context, client k8sclient.Client, namespace string) {
+	redis := &crov1alpha1.Redis{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "throw-away-redis-pod",
+			Namespace: namespace,
+		},
+	}
+
+	if err := client.Delete(ctx, redis); err != nil && !k8serr.IsNotFound(err) {
+		fmt.Printf("Failed to clean up throw away redis: %v. Please delete manually:\n", err)
+		fmt.Println("Run: oc delete redis throw-away-redis-pod -n redhat-rhoam-operator")
+	}
+}
+
+// testCountIncreases performs numberOfRequests simultaneously and asserts that
+// the counter in Redis is increased at least that amount of times. Returns
+// true with no error if the test succeeds. false with no error if the test failed
+// and an error if an unexpected error occurred when performing the test
+func testCountIncreases(r *redisCounter, endpoint string, numberOfRequests int) (bool, error) {
+	if err := r.Flush(); err != nil {
```

Comment:
hey @sergioifg94 the test looks brilliant!

I've just a question around flushing redis - if we are flushing before each series of pings, do the "previous count" matter - in my understanding from looking at the code - it would be equivalent to 0 every time. 
Or, we could get the current count before flushing? as it seems that the GetCount will always get 0 value?

Reply:
Thanks Michal, I initially hadn't added the flush call but it was failing at random when calling `GetCount`. The flush appeared to make it reliable but yes, after the flush it would always be 0

Reply:
I think the reason why it was unreliable is we are keeping track not only of the current minute but few past minutes too, where each minute has a new key-value pair in redis. I see you are filtering the liveness probe from the redis call but there's also the "old" minutes and since we are getting the values back by the "*", we are getting the results of current + couple last minutes - flush clears it so there's always only 1 minute. 
The issue I could see we could possibly have, is around if the tests are run between 2 minutes. For example, if you start hitting an enpoint at 00:59 and finish at 01:00 - there will be 2 ratelimit counts and a liveness probe.
The workaround this in manual testing was to flush every 1min mark or, when there's more than 2 entries in redis.